### ISSUE TITLE
Use-after-free/double-remove on mmu_interval_notifier in amdxdna kernel driver leads to RCU stalls and kernel lockup

### ISSUE THREAD

**Issue by lgarbarini** [amd/xdna-driver#595]
Moving from gfx1151 pytorch bringup thread. More details are there including full reproduction details. [[original discussion on rocm/TheRock](https://github.com/ROCm/TheRock/discussions/655#discussioncomment-13440258)].

I think I may have found a bug triggered during multi-threaded application exit or VMA teardown. The bug occurs when multiple amdxdna_gem_vm_close() calls target BOs or VMAs sharing the same mmu_interval_notifier, leading to double removal and kernel memory corruption, spinlock hangs, and RCU stalls.

System information
- Kernel: 6.14.9-300.fc42.x86_64
- Hardware: HP ZBook Ultra G1a (14 inch Mobile Workstation)
- Module: amdxdna
- Userspace reproducer: Multi-threaded workload with rapid open/close or exit on buffers mapped with mmap

Multiple mmu_interval_notifier_remove calls are made in quick succession with the same (shared) notifier pointer but from different BO/VM object pointers and different PIDs. I'm new to this driver, but I suspect it is internal to the driver as userspace cannot directly expose or pass in notifier pointers as far as I can tell.

Using eBPF, I think I have an example of the problem. I do not believe `mmu_interval_notifier_remove` should ever be called with the same pointer multiple times.
```
amdxdna_gem_vm_close obj 0xffff888a949fcdd8 pid 6787 ...
mmu_interval_notifier_remove arg 0xffff8889107039f8 pid 6787 ...

amdxdna_gem_vm_close obj 0xffff888b0a87aa98 pid 6781 ...
mmu_interval_notifier_remove arg 0xffff8889107039f8 pid 6781 ...

amdxdna_gem_vm_close obj 0xffff888ab1ca2ea8 pid 6785 ...
mmu_interval_notifier_remove arg 0xffff8889107039f8 pid 6785 ...
...

[eBPF_output.log](https://github.com/user-attachments/files/20744744/eBPF_output.log)
[trace_lockup.bt.txt](https://github.com/user-attachments/files/20744743/trace_lockup.bt.txt)

**Comment by lgarbarini** [comment]
Here's a minimal reproducible example I could come up with in pytorch:

```python
import numpy as np
import torch
from torch.utils.data import Dataset, DataLoader
import multiprocessing as mp
import os, time, random

class SharedNpyDataset(Dataset):
    def __init__(self):
        self.arr = (np.random.randn(8,32,64,64)*3).astype(np.float32)
    def __len__(self):
        return 64
    def __getitem__(self, idx):
        return self.arr

def worker_init(worker_id):
    print(f"DATA WORKER {worker_id} pid={os.getpid()} started.")

def launch_wave(n_workers=12, n_reps=8):
    dataset = SharedNpyDataset()
    loader = DataLoader(dataset, batch_size=1, shuffle=True, 
        num_workers=n_workers, pin_memory=True, worker_init_fn=worker_init)
    model = torch.nn.Conv3d(8, 2, 3, padding=1).eval().to('cuda')
    cnt = 0
    for batch in loader:
        batch = batch.float().to('cuda', non_blocking=True)
        with torch.no_grad():
            out = model(batch)
        x = out.cpu().numpy()
        del batch, out, x
        torch.cuda.synchronize()
        time.sleep(random.uniform(0, 0.05))
        cnt += 1
        if cnt > n_reps: break
    del loader, dataset

if __name__ == "__main__":
    for i in range(24):
        print("CRASH TEST WAVE", i)
        launch_wave(12, 12)
        time.sleep(random.uniform(0.1, 0.2))
```

**Comment by lgarbarini** [comment]
I have an even simpler C++ MRE that can trigger the error. Note that the child processes don't even need to use device calls, just having mapped memory and having instantiated the device is enough. 

```cpp
#include <chrono>
#include <fcntl.h>
#include <fstream>
#include <hip/hip_runtime.h>
#include <iostream>
#include <sys/mman.h>
#include <sys/wait.h>
#include <thread>
#include <unistd.h>
#include <vector>

#define NPROCS 8
#define NITERS 10
#define FILE_SIZE (8 * 32 * 64 * 16) // 32 MB

#define HIP_CHECK(call)                                                        \
  do {                                                                         \
    hipError_t _err = (call);                                                  \
    if (_err != hipSuccess) {                                                  \
      std::cerr << "HIP error " << __FILE__ << ":" << __LINE__ << " "          \
                << hipGetErrorString(_err) << std::endl;                       \
      _exit(1);                                                                \
    }                                                                          \
  } while (0)

void stress_child(int idx, float *file_map, size_t n_elts) {
  float sum = 0.0f;
  // Just read portions of file_map
  for (int it = 0; it < NITERS; ++it) {
    for (size_t i = idx * 1024; i < idx * 1024 + 256 && i < n_elts; ++i)
      sum += file_map[i];
    std::this_thread::sleep_for(std::chrono::milliseconds(5 + rand() % 20));
  }
  _exit(0);
}

int main() {
  // 1. Prepare file
  std::string fname = "/tmp/mremaptest.dat";
  std::ofstream f(fname, std::ios::binary | std::ios::trunc);
  size_t n_elts = FILE_SIZE;
  std::vector<float> zeros(n_elts, 0.0f);
  for (size_t i = 0; i < n_elts; ++i)
    zeros[i] = (float)i;
  f.write(reinterpret_cast<const char *>(zeros.data()), n_elts * sizeof(float));
  f.close();

  // 2. mmap file
  int fd = open(fname.c_str(), O_RDONLY);
  void *map = mmap(NULL, n_elts * sizeof(float), PROT_READ, MAP_SHARED, fd, 0);
  if (map == MAP_FAILED) {
    std::cerr << "mmap failed" << std::endl;
    return 1;
  }
  float *file_map = reinterpret_cast<float *>(map);

  HIP_CHECK(hipSetDevice(0));
  // 3. Fork
  for (int round = 0; round < 10; ++round) {
    std::vector<pid_t> children;
    for (int i = 0; i < NPROCS; ++i) {
      pid_t pid = fork();
      if (pid == 0) {
        stress_child(i, file_map, n_elts);
      }
      children.push_back(pid);
      usleep(rand() % 12000);
    }
    for (auto pid : children)
      waitpid(pid, NULL, 0);
    std::cout << "wave " << round << " finished\n";
    std::this_thread::sleep_for(std::chrono::milliseconds(20 + rand() % 100));
  }

  munmap(map, n_elts * sizeof(float));
  close(fd);
  std::cout << "All waves done.\n";
  return 0;
}
```

[dmesg_trace.log](https://github.com/user-attachments/files/20747311/dmesg_trace.log)
[eBPF_output.log](https://github.com/user-attachments/files/20747310/eBPF_output.log)
[trace_lockup.bt.txt](https://github.com/user-attachments/files/20747313/trace_lockup.bt.txt)

**Comment by lgarbarini** [comment]
Now that I actually read through the suspect code, it appears to be resolved in mainline > 6.15 by https://lore.kernel.org/all/20250325200105.2744079-1-lizhi.hou@amd.com/ and in this repo here: https://github.com/amd/xdna-driver/commit/49c590dc74fe5779bf613e357ed0a57efba0b4df
